src/vsr/journal.zig

```diff
@@ -571,15 +703,17 @@ pub fn Journal(comptime Replica: type, comptime Storage: type) type {
                 return;
             }
 
-            _ = replica.journal.entry_for_op_exact_with_checksum(op, checksum) orelse {
-                self.read_prepare_log(op, checksum, "no entry exactly");
-                read.callback(replica, null, null);
-                return;
-            };
+            // Check that the `headers` slot belongs to the same op that it did when the read began.
+            // The slot may not match the Read's op/checksum due to either:
+            // * The in-memory header changed since the read began.
+            // * The in-memory header is reserved+faulty; the read was via `prepare_checksums`
+            const slot = self.slot_with_op_and_checksum(op, checksum);
 
             if (!read.message.header.valid_checksum()) {
-                self.faulty.set(op);
-                self.dirty.set(op);
+                if (slot) |s| {
+                    self.faulty.set(s);
+                    self.dirty.set(s);
+                }
 
```

Comment:
If the read failed because of a checksum mismatch, and if the slot is supposed to still contain the same op/checksum, then should we clear the checksum in `prepare_checksums` so it can't be used to generate further I/O load when we know further reads for the slot will probably fail again until it's repaired?

Reply:
> You'll see I also added a TODO to use a `?u128` type for `prepare_checksums`. This will use a little more memory but over 1024 slots not that much, and it will be cleaner to use null as an explicit sentinel rather than repurpose 0, as that is still a valid checksum value.


`request_prepare` uses `checksum=0` to indicate "any checksum" — doesn't that mean that `0` _can't_ be a valid checksum? A leader with a corrupt message would have no way of recovering a `checksum=0` message from its followers (they'd ignore the request).

Though I guess the view change would time out and eventually the replica with the valid `checksum=0` message would be leader, and then it could progress. Is that what you mean?

Reply:
> If the read failed because of a checksum mismatch, and if the slot is supposed to still contain the same op/checksum, then should we clear the checksum in `prepare_checksums` so it can't be used to generate further I/O load when we know further reads for the slot will probably fail again until it's repaired?

I'd rather leave it intact. It is possible that the checksum mismatch is only due to a misdirected read, and in reality the prepare is intact on disk. i.e. a retry may actually succeed.

Reply:
> request_prepare uses checksum=0 to indicate "any checksum" — doesn't that mean that 0 can't be a valid checksum? A leader with a corrupt message would have no way of recovering a checksum=0 message from its followers (they'd ignore the request).

Ah, I had forgotten about this. Yes, it does mean that `0` can't be a valid checksum, unless we fix how we communicate with `request_prepare()` or unless we `OR` in a `1` when we calculate the checksum.

> Though I guess the view change would time out and eventually the replica with the valid checksum=0 message would be leader, and then it could progress. Is that what you mean?

No, I hadn't thought ahead this far! 🙂